framework/DataObjects/DataSet.py

```diff
@@ -0,0 +1,1853 @@
+      else:
+        res = self._data[var]
+    elif isinstance(var,list):
+      res = dict((v,self.getVarValues(v)) for v in var)
+    else:
+      self.raiseAnError(RuntimeError,'Unrecognized request type:',type(var))
+    return res
+
+  def load(self,fileName,style='netCDF',**kwargs):
+    """
+      Reads this dataset from disk based on the format.
+      @ In, fileName, str, path and name of file to read
+      @ In, style, str, optional, options are enumerated below
+      @ In, kwargs, dict, optional, additional arguments to pass to reading function
+      @ Out, None
+    """
+    style = style.lower()
+    # if fileToLoad in kwargs, then filename is actualle fileName/fileToLoad
+    if 'fileToLoad' in kwargs.keys():
+      fileName = kwargs['fileToLoad'].getAbsFile()
+    # load based on style for loading
+    if style == 'netcdf':
+      self._fromNetCDF(fileName,**kwargs)
+    elif style == 'csv':
+      # make sure we don't include the "csv"
+      if fileName.endswith('.csv'):
+        fileName = fileName[:-4]
+      self._fromCSV(fileName,**kwargs)
+    elif style == 'dict':
+      self._fromDict(fileName,**kwargs)
+    # TODO dask
+    else:
+      self.raiseAnError(NotImplementedError,'Unrecognized read style: "{}"'.format(style))
+    # after loading, set or reset scaling factors
+    self._setScalingFactors()
+
+  def realization(self,index=None,matchDict=None,tol=1e-15, unpackXArray=False):
+    """
+      Method to obtain a realization from the data, either by index or matching value.
+      Either "index" or "matchDict" must be supplied.
+      If matchDict and no match is found, will return (len(self),None) after the pattern of numpy, scipy
+      @ In, index, int, optional, number of row to retrieve (by index, not be "sample")
+      @ In, matchDict, dict, optional, {key:val} to search for matches
+      @ In, tol, float, optional, tolerance to which match should be made
+      @ In, unpackXArray, bool, optional, True if the coordinates of the xarray variables must be exposed in the dict (e.g. if P(t) => {P:ndarray, t:ndarray}) (valid only for dataset)
+      @ Out, index, int, optional, index where found (or len(self) if not found), only returned if matchDict
+      @ Out, rlz, dict, realization requested (None if not found)
+    """
+    # TODO convert input space to KD tree for faster searching -> XArray.DataArray has this built in?
+    ## first, check that some direction was given, either an index or a match to find
+    if (index is None and matchDict is None) or (index is not None and matchDict is not None):
+      self.raiseAnError(TypeError,'Either "index" OR "matchDict" (not both) must be specified to use "realization!"')
+    numInData = len(self._data[self.sampleTag]) if self._data is not None else 0
+    numInCollector = len(self._collector) if self._collector is not None else 0
+    ## next, depends on if we're doing an index lookup or a realization match
+    if index is not None:
```

Comment:
Should index lookup be broken out into a [] operator?

Reply:
We're replicating that behavior, but we can't get away with just the square brackets because we need to pass additional arguments to the "getter".  The underlying data structure is not so simple as that, unfortunately.